#1778 Port ColorsLab to WPF

blewjy opened the pull request:
Fixes #1778. Ready for review.

~As mentioned in the issue, this PR has the proposed functionality for the buttons, which is where user selects the colour before clicking the button to apply the color to the shape. This feature can be easily changed back to the old behaviour if that is desired.~

This PR will port ColorsLab from the WindowsForms to the WPF so as to keep the implementation consistent throughout the labs.

- [x] Base pane and UI in WPF
- [x] Main selected color rectangle & color picker
- [x] Button functionality (click-and-drag to eyedrop color)
- [x] Sliders + matching colors
- [x] Drag-and-drop to favourite colors
- [x] Save, load, clear & reload favourtie colors
- [x] Eyedropper tool
- [x] Rewrite tests
- [x] Color Information Dialog

This new implementation of ColorsLab follows the old implementation very closely, with some minor exceptions. 

**Previous implementation:** To change color of the main rectangle, either click to open color picker dialog or click and drag to eyedrop.
**New implementation:** Instead of dragging the rectangle to eyedrop, there is now a new eyedropper button on the top row. This allows user to drag the main color rectangle to the favourite colors panel as well.

Aside from a couple of bug fixes, other main features remain the same.

blewjy commented:
> Actually, there is an alternative we can consider, which might alleviate the unintuitiveness of the feature UX.
We can lock into 'magnifying glass + live preview' mode upon clicking on the button (instead of requiring to click and drag). We can go back to normal mode upon clicking the button again, or picking a color, or pressing `ESC` key.

From the issue, there was a suggestion from Prof to change the button functionality to a click to "lock-into-mode" and click again to select, instead of the initial drag-to-pick color. 

After much research and testing, I found that while allowing the user to click to go into the color picking mode was more intuitive, it might not be a feasible feature to implement, at least not without some hacky solutions. This is because after the user locks into that mode, clicking on a location outside of PowerPoint causes PowerPoint to lose focus. While the colour can still be successfully picked, it makes for a pretty bad experience for the user. 

As such, I will be reverting back to the old behaviour for now, which is the click drag and release behaviour. Releasing mouse button outside of PowerPoint does not cause it to lose focus, hence this method works.

ChesterSng commented:
Hi @blewjy great job on this PR! Colors Lab FT passing on PowerPoint 2016. 

Putting the title of the icons in the colorslab pane will help users understand the icon quicker, instead of using the tooltip only.

Found a few bugs:  

**1.) Error when dragging the color from color box into textbox**
_Steps to reproduce:_
1. Select and drag from the colorpicker (or any color box)
2. Release inside any textbox (even shapes that have textbox)
![catastrophic failure](https://user-images.githubusercontent.com/39242483/53073499-26a53700-3523-11e9-94a2-7569be91819d.gif)

**2.) Main colour changes even when eye dropper tool is not used properly**
_Steps to reproduce:__
1. Change the main color.
2. Click on the `Eye Dropper Tool`, don't drag.
3. Error appears but main colour changed.

_Main color will be changed to the last color that was applied to it successfully by the eyedropper tool_

Compared to the behaviour of other tools, they do not change the colour of the text/shape/line when there is an error. 
![main colour change with error](https://user-images.githubusercontent.com/39242483/53074916-efd12000-3526-11e9-8bc6-a3312669769a.gif)

blewjy commented:
Hey @ChesterSng, thanks for testing!

**1.) Error when dragging the color from color box into textbox**
It seems that this error is due to PowerPoint not having the proper support for such drag and drop operations from an external source into its textbox/shapes. I got around this by catching the exception thrown and inserting it into the `TextRange` of the textbox/shape instead. [Link for more info](https://social.msdn.microsoft.com/Forums/vstudio/en-US/16fbc569-fd8b-4447-8d4e-c524954dcb78/wpf-listbox-drop-and-powerpoint?forum=wpf)

**2.) Main colour changes even when eye dropper tool is not used properly**
Fixed.

I also added labels to the buttons as suggested. Thanks!

YuPeiHenry commented:
@blewjy Tested and working on PPT2010 (including FT). There appears to be a minor issue with save/load on favorite colors. The file extension is unspecified.

Original:
![image](https://user-images.githubusercontent.com/40821389/53156459-69860e00-35fa-11e9-882a-cbac7fc21745.png)

Current:
![image](https://user-images.githubusercontent.com/40821389/53156636-c8e41e00-35fa-11e9-8a1f-d47fb22fc770.png)

blewjy commented:
@YuPeiHenry good catch! Fixed it!

YuPeiHenry commented on PowerPointLabs/PowerPointLabs/DataSources/ColorDataSource.cs:
```diff
@@ -259,7 +259,61 @@ public HSLColor ThemeColorTen
+                if (value != this.themeColorElevenValue)
+                {
+                    this.themeColorElevenValue = value;
+                    OnPropertyChanged("themeColorEleven");
```
Seems like there are a lot of colors. I'm not sure if its possible (and not too difficult) to change this to use an array, but it would be good.

blewjy replied:
This `ColorDataSource.cs` is mostly code from the old implementation of ColorsLab which i reused. Because all of them are using bindings with the UI and calling that `OnPropertyChanged` method, I don't think using a simple array will work for individual objects, i.e. they won't get updated as expected. 

One possible solution we can look into is using `ObservableCollections`, however I have not researched much into this just yet. Perhaps this could be something to consider refactoring in the near future, but as of now I don't think this affects functionality much!

blewjy commented on PowerPointLabs/PowerPointLabs/ActionFramework/ColorsLab/ColorsLabActionHandler.cs:
```diff
@@ -12,11 +13,17 @@ class ColorsLabActionHandler : ActionHandler
+            // if currently the pane is hidden, show the pane
+            if (!colorPane.Visible)
             {
-                colorPane.Visible = !colorPane.Visible;
```
This is required to hide and show the colorPane when user clicks on the main ribbon button

blewjy replied:
No difference... I think I took this from one of the other labs. I've shortened this block of code.

blewjy commented on PowerPointLabs/PowerPointLabs/ColorsLab/HSLColor.cs:
```diff
@@ -44,6 +44,12 @@ public string ToRGBString()
+        public string ToHEXString()
+        {
+            Color color = (Color)this;
+            return "#" + color.R.ToString("X2") + color.G.ToString("X2") + color.B.ToString("X2");
```
Good idea! I have refactored this into `ColorHelper.cs`. Makes more sense this way, thanks for the suggestion!

leeyh20 replied:
Instead of still using `ToString("X2")`

blewjy commented:
Tested and working on PPT2013. This PR is ready for mentors' review.

junkiattan commented:
@alexfjw @leeyh20 

Code LGTM. On my end, I have not touched WPF to the extent of this PR.
Great initiative to maintain consistency with the labs though. Im sure this would make future developments alot more easier to code and maintain;)

leeyh20 commented on PowerPointLabs/PowerPointLabs/Converters/ColorPane/SelectedColorShiftLuminosityByFactor.cs:
```diff
@@ -0,0 +1,32 @@
+        {
+            if (value == null)
+            {
+                Color color = (HSLColor)value;
```
Was wondering if `value == null` then converting it to `HSLColor` wouldn't it be still null? Or is there a default value to the `HSLColor`?

blewjy replied:
That's right, `HSLColor` when implicitly converted to `System.Drawing.Color` will handle `null` by setting the `Color` to `rgb(0, 0, 0)`. 

`Color` is a struct in C#, so not allowed to set as `null`. If I'm not wrong, there will be a `NullReferenceException` thrown if we attempt to set `Color` to `null`. But since we have our own `HSLColor`, this is handled here.

leeyh20 commented on PowerPointLabs/PowerPointLabs/FunctionalTestInterface.Impl/Controller/ColorsLabController.cs:
```diff
@@ -16,114 +23,364 @@ class ColorsLabController : MarshalByRefObject, IColorsLabController
+                    rect.RaiseEvent(
+                        new MouseButtonEventArgs(Mouse.PrimaryDevice, Environment.TickCount, MouseButton.Left)
+                        {
+                            RoutedEvent = UIElement.MouseDownEvent
```
This routed event code seems to be repeated quite a few times, would it be possible to refactor into a function?

blewjy replied:
Good idea, done!

leeyh20 commented on PowerPointLabs/PowerPointLabs/TextCollection/ColorsLabText.cs:
```diff
@@ -14,17 +14,16 @@ internal static class ColorsLabText
 
         public const string MainColorBoxTooltip = "Choose the main color: " +
-                                   "\r\nDrag the button to pick a color, " +
+                                   "\r\nYou can drag the box to favorites palatte, " +
```
Maybe start the sentence with "Drag the box..." so that it is more consistent?

blewjy replied:
Done

blewjy commented on PowerPointLabs/PowerPointLabs/ColorsLab/HSLColor.cs:
```diff
@@ -44,6 +44,12 @@ public string ToRGBString()
             return String.Format("R: {0:#0.##} G: {1:#0.##} B: {2:#0.##}", color.R, color.G, color.B);
         }
 
+        public string ToHEXString()
```
Oops, changed!

leeyh20 commented:
Thanks for this PR! 👍
